{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### WesterLab scRNA_seq analysis\n",
    "\n",
    "Tools for downstream analysis of scRNA-seq differential expression data\n",
    "\n",
    "Subroutines created by Anthony Moussa and presented in Moussa and Wester, 2022 https://doi.org/10.1101/2022.06.14.496156"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importing libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function Name\n",
    "\n",
    "description: parameters and result\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "body of text for code with comments for key algorithms as would be described in pseudocode"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GOPrep\n",
    "pass a dataframe with compiled DE data to be labeled and filtered for GO terms\n",
    "\n",
    "df = combined_df\n",
    "\n",
    "FinalGO = last GO term in combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GOPrep(df, FinalGO):\n",
    "    \n",
    "    # label GO relevant genes\n",
    "    df.loc[df['Gene'].str.lower().isin(df[\"CCA.1\"].str.lower()), 'CCA'] = 1\n",
    "\n",
    "    df.loc[df['Gene'].str.lower().isin(df[\"RCCA.1\"].str.lower()), 'RCCA'] = 1\n",
    "\n",
    "    df.loc[df['Gene'].str.lower().isin(df[\"RTSS.1\"].str.lower()), 'RTSS'] = 1\n",
    "\n",
    "    # select GO relevant genes\n",
    "    while True:\n",
    "        filtered = input(\"Labeled! Would you like to filter out non-GO genes and non-DE genes (y/n):\")\n",
    "        if filtered == 'y':\n",
    "            df = df.loc[(df['CCA'] == 1) | (df['RCCA'] == 1) | (df['RTSS'] == 1)]\n",
    "            df = df.loc[~(df['p_val_adj'] > .05)]\n",
    "            break\n",
    "        elif filtered == 'n':\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "    \n",
    "    GOPrep = df\n",
    "    \n",
    "    cutoff = GOPrep.columns.get_loc(FinalGO)+1\n",
    "   \n",
    "    # drop remnant columns from filtering\n",
    "    GOPrep.drop(GOPrep.columns[cutoff:], axis=1, inplace=True)\n",
    "    return GOPrep"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subsetter\n",
    "pass a dataframe to be subsetted and sorted\n",
    "\n",
    "df = combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subsetter(df):\n",
    "\n",
    "    total_df = df\n",
    "    \n",
    "    #subset for GO terms\n",
    "    while True:\n",
    "        GO = input(\"Subset for all GOs (all), Cell-Cell Adhesion (CCA), Regulation of CCA (RCCA), Regulation of Trans-Synaptic signaling (RTSS): \")\n",
    "        if GO == \"all\":\n",
    "            break\n",
    "        elif GO == \"CCA\" or GO == \"RCCA\" or GO == \"RTSS\":\n",
    "            total_df = total_df.loc[(total_df[GO] == 1)]\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "\n",
    "    #subset for cell classes\n",
    "    while True:\n",
    "        cellclass = input(\"Subset for all data (all), glutamatergic (Glut) or GABAergic (GABA): \")\n",
    "        if cellclass == \"all\":\n",
    "            break\n",
    "        elif cellclass == \"Glut\" or cellclass == \"GABA\":\n",
    "            total_df = total_df.loc[total_df[\"Class\"] == cellclass]\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "\n",
    "    #subset for brain regions\n",
    "    while True:\n",
    "        brain_region = input(\"Subset for all data (all), ALM vs VISp, ALM, VISp, or combination (combo): \")\n",
    "        if brain_region == \"all\":\n",
    "            break\n",
    "        elif brain_region == \"ALM vs VISp\" or brain_region == \"ALM\" or brain_region == \"VISp\":\n",
    "            total_df = total_df.loc[total_df[\"brain_region\"] == brain_region]\n",
    "            break\n",
    "        elif brain_region == \"combo\":\n",
    "            while True:\n",
    "                Regions = input(f\"Available brain_regions: {list(df['brain_region'].unique())} \\n\"\n",
    "                               \"Please (re)enter list of brain_regions from those available (comma separated values, no spaces after commas):  \")\n",
    "                more = \"n\"\n",
    "                for x in Regions.split(\",\"):\n",
    "                    if not (x in list(df['brain_region'].unique())):\n",
    "                                 print(f\"{x} is not an available brain region\")\n",
    "                                 more = \"y\"\n",
    "                                 L=[]\n",
    "\n",
    "                if more == \"n\":\n",
    "                    more = input(f\"Content with current brain regions:{Regions.split(',')} \\n\" \n",
    "                                 \"(y/n) (n restarts function): \")\n",
    "                    if more != \"n\":\n",
    "                        total_df = total_df.loc[df['brain_region'].isin(Regions.split(','))]\n",
    "                        break\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "    \n",
    "    #subset for subclasses\n",
    "    if input(\"Subset for specific subclasses (y/n) (y = specifics; n = all subclasses): \") == \"y\":\n",
    "        while True:\n",
    "            SubClasses = input(f\"Available subclasses: {list(total_df['Subclass'].unique())} \\n\"\n",
    "                           \"Please (re)enter list of subclasses from those available (comma separated values, no spaces after commas):  \")\n",
    "            more = \"n\"\n",
    "            for x in SubClasses.split(\",\"):\n",
    "                if not (x in list(total_df['Subclass'].unique())):\n",
    "                             print(f\"{x} is not an available subclass\")\n",
    "                             more = \"y\"\n",
    "\n",
    "            if more == \"n\":\n",
    "                more = input(f\"Content with current subclasses:{SubClasses.split(',')} \\n\" \n",
    "                             \"(y/n) (n restarts function): \")\n",
    "                if more != \"n\":\n",
    "                    total_df = total_df.loc[total_df['Subclass'].isin(SubClasses.split(\",\"))]\n",
    "                    break\n",
    "    \n",
    "    #sorter for plotting\n",
    "    if input(\"Would you like to sort subsetted data? (y/n) \") == \"y\":\n",
    "        while True:        \n",
    "            Sort = input(\"Sort by GABAergic/Glutmatergic (C), brain region (R), or subclass (SC) (list order, comma separated): \")\n",
    "\n",
    "            D = {\"C\": \"Class\", \"R\": \"brain_region\", \"SC\": \"subclass\"}\n",
    "\n",
    "            more = \"n\"\n",
    "            L = Sort.split(\",\")\n",
    "            for x in L:\n",
    "                if not (x in [\"C\",\"R\",\"SC\"]):\n",
    "                             print(f\"{x} is not an available sort parameter\")\n",
    "                             more = \"y\"\n",
    "\n",
    "            if more == \"n\":\n",
    "                more = input(f\"Content with current subclasses:{Sort.split(',')} \\n\" \n",
    "                             \"(y/n) (n restarts function): \")\n",
    "\n",
    "                if more != \"n\":\n",
    "                    total_df = total_df.sort_values([D[L[0]],D[L[1]], D[L[2]]])\n",
    "                    break     \n",
    "                                                       \n",
    "    return GO, cellclass, brain_region, total_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### clinExploder\n",
    "pass a clinical df (e.g. uploaded SFARI ASD list) with one or more columns with csv strings and will use Pandas Explode function to split/explode such columns into separate rows.\n",
    "\n",
    "e.g break up row with multiple clinical descriptions into multiple rows\n",
    "\n",
    "preset for \"conditions\" column\n",
    "\n",
    "df_clin - used passed clinical df. e.g. SFARI list with columns for genes and ASD classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clinExploder(df_clin):\n",
    "    \n",
    "    while True:\n",
    "        SFARI = input(\"Are you passing the ASD SFARI gene list with 'Condition' and 'syndromic' columns? (y/n):\" )\n",
    "        if SFARI == \"y\":\n",
    "            df_clin.loc[df_clin['Condition'] == 3.0, 'Condition'] = \"Suggestive Evidence\"\n",
    "            df_clin.loc[df_clin['Condition'] == 2.0, 'Condition'] = \"Strong Candidate\"\n",
    "            df_clin.loc[df_clin['Condition'] == 1.0, 'Condition'] = \"High Confidence\"\n",
    "            df_clin.loc[df_clin['Condition'] == np.nan, 'Condition'] = \"\"\n",
    "            df_clin.loc[df_clin['syndromic'] == 1.0, 'syndromic'] = \"Syndromic\"\n",
    "            df_clin.loc[df_clin['syndromic'] == 0, 'syndromic'] = \"\"\n",
    "            df_clin.loc[df_clin['Condition'].isnull(), 'Condition'] = \"\"\n",
    "            df_clin.loc[(df_clin['Condition'] != \"\") & (df_clin['syndromic'] == 'Syndromic'), 'Conditions'] = df_clin[\"Condition\"] + \",\" + df_clin[\"syndromic\"]\n",
    "            df_clin.loc[(df_clin['Condition'] == \"\") & (df_clin['syndromic'] == 'Syndromic'), 'Conditions'] = df_clin[\"syndromic\"]\n",
    "            df_clin.loc[(df_clin['syndromic'] == \"\"), 'Conditions'] = df_clin[\"Condition\"]\n",
    "            break\n",
    "        elif SFARI == 'n':\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "       \n",
    "    df_clin = df_clin[['Gene','Conditions']]\n",
    "\n",
    "    df_clin['Conditions'] = df_clin['Conditions'].apply(lambda x: x.split(','))\n",
    "\n",
    "    df_clin = df_clin.explode('Conditions')\n",
    "\n",
    "    df_clin.columns = ['Gene','Condition']\n",
    "\n",
    "    df_clin = df_clin.reset_index(drop =True)\n",
    "    \n",
    "    return df_clin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ClinPrep\n",
    "\n",
    "prepares a filtered clinical df for all downstream clinical functions\n",
    "\n",
    "df_clin = df that captures all potential clinical categories (can adjust based on stringency of clinical classfication); columns required: Gene, Condition\n",
    "\n",
    "df_DEGenes = df with all possible DE Genes; columns required: DEGlut, DEGABA, DE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ClinPrep(df_clin,df_DEGenes):\n",
    "      \n",
    "    # Break up rows with comma delimits\n",
    "    df_clin = pd.concat([pd.Series(row['Condition'], row['Gene'].split(', '))              \n",
    "                        for _, row in df_clin.iterrows()]).reset_index()\n",
    "\n",
    "    df_clin.columns = ['Gene', 'Conditions']\n",
    "\n",
    "    df_clin = df_clin.drop_duplicates(subset=['Gene', 'Conditions'], keep ='first')\n",
    "\n",
    "    # Filter for clinically relevant DE genes; most important part of function\n",
    "    df_clin = df_clin.loc[df_clin[\"Gene\"].str.lower().isin(df_DEGenes[\"DE\"].str.lower())]\n",
    "\n",
    "    # label for gene origin; gives high level overview of Glutamatergic/GABAergic clinical contributions\n",
    "    df_clin[\"Class\"] = 0\n",
    "    df_clin.loc[df_clin[\"Gene\"].str.lower().isin(df_DEGenes[\"DE Glut\"].str.lower()), \"Class\"] = 1\n",
    "    df_clin.loc[df_clin[\"Gene\"].str.lower().isin(df_DEGenes[\"DE GABA\"].str.lower()), \"Class\"] = df_clin[\"Class\"]-1\n",
    "    df_ClinPrep = df_clin\n",
    "    \n",
    "    return df_ClinPrep"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ClinicalFilter\n",
    "\n",
    "\"select clinically relevant genes of interest\" from ClinPrep output - i.e. subsets ClinPrep for only genes relevant to conditions in narrowed list; will remove condition labels for gene that are clinically relevant but not in provided dictionary\n",
    "\n",
    "df = output of ClinPrep\n",
    "\n",
    "df_ClinList = user passed clinical df that is converted into dictionary and informs which conditions to keep and how they are labeled **relies on preset ConditionList that is manually made"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ClinicalFilter (df,df_ClinList):\n",
    "    \n",
    "    clinical = df_ClinList\n",
    "\n",
    "    d = clinical.groupby('Narrow')['Conditions'].apply(list).to_dict()\n",
    "    \n",
    "    df['Found'] = 0\n",
    "    \n",
    "#   annotates genes found in dictionary: key = user meta condition (Neurodevelopmental), value = database condition name\n",
    "    for key, value in d.items():\n",
    "        df.loc[df[\"Conditions\"].isin(value), [\"Conditions\", 'Found']] = [key,1]\n",
    "\n",
    "#   offers option to remove genes not found in dictionary\n",
    "    while True:\n",
    "        remove = input(\"Would you like to remove genes not found associated with a condition in your condition dictionary but still clinically relevant? (y/n):\")\n",
    "        if remove == 'y':\n",
    "            for key, value in d.items():\n",
    "                df.loc[df['Found'] != 1, ['Gene',\"Conditions\"]] = ['NAN','NAN']\n",
    "            break\n",
    "        elif remove == 'n':\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "    \n",
    "    df = df.drop_duplicates(subset=['Gene', 'Conditions'], keep ='first')\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ClinLabel\n",
    "\n",
    "\"label DE genes by selected clinically relevant genes\" from output of ClinPrep\n",
    "\n",
    "takes a df_clinical and df_DE and filters then labels DE genes which are clinically relevant\n",
    "\n",
    "df_clinical = output from ClinPrep\n",
    "\n",
    "df_DE = user passed df of all DE genes\n",
    "\n",
    "df_ClinList = user passed df of all conditions of interest for analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#labels gene by conditions (comma separated strings)\n",
    "\n",
    "def ClinLabel(df_clinical,df_DE,df_ClinList):\n",
    "    \n",
    "    df_clinical = ClinicalFilter(df_clinical,df_ClinList)\n",
    "\n",
    "    #Offers option to annotate all of dataset or just clinically relevant DE genes\n",
    "    while True: \n",
    "        remove = input(\"Would you like to remove non-clinically relevant DE genes (y/n): \")\n",
    "        if (remove == 'y'):\n",
    "            df_DE = df_DE.loc[df_DE[\"Gene\"].str.lower().isin(df_clinical[\"Gene\"].str.lower())]\n",
    "            df_DE = df_DE.reset_index(drop=True)\n",
    "            break\n",
    "        elif remove == 'n':\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "\n",
    "    for gene in df_DE[\"Gene\"]:\n",
    "        clinical_gene = df_clinical.loc[df_clinical['Gene'].str.lower() == gene.lower()]\n",
    "        #solves issue of ASD/Neurodev vs Neurodev/ASD\n",
    "        cList = sorted(clinical_gene['Conditions'].tolist())\n",
    "        my_string = ','.join(cList)\n",
    "        df_DE.loc[df_DE['Gene'] == gene, \"Conditions\"] = my_string\n",
    "\n",
    "    return df_DE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DECollapser\n",
    "takes a filtered and labeled df and gives DE genes, noting subclass origin (ALM-specific, VISp-specific, conserved, unique)\n",
    "\n",
    "Collapser designed for level 3 granularity (brain region and subclass 1 vs 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DECollapser(df):\n",
    "    \n",
    "    result = subsetter(df)\n",
    "    GO = result[0]\n",
    "    cellclass = result[1]\n",
    "    brain_region = result[2] \n",
    "    df = result[3]\n",
    "\n",
    "    comparisons = df['Comparison'].unique()\n",
    "\n",
    "    while True:\n",
    "        clin = input('Are you passing in a df labeled with conditions? (y/n): ')\n",
    "        if clin == \"y\":\n",
    "            break\n",
    "        elif clin == \"n\":\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "\n",
    "    rows = []\n",
    "    \n",
    "#     create empty df to store DE test data that is different between brain regions\n",
    "    df_unconserved = df.loc[df['Class'] == 0]\n",
    "    \n",
    "    for comparison1 in comparisons:\n",
    "            df1 = df.loc[df[\"Comparison\"] == comparison1]\n",
    "            df1 = df1.reset_index(drop=True)\n",
    "            for comparison2 in comparisons:\n",
    "                    df2 = df.loc[df[\"Comparison\"] == comparison2]\n",
    "                    df2 = df2.reset_index(drop=True)\n",
    "                    \n",
    "        #             indicates we are doing a same subclass comparison\n",
    "                    if (df1['Comparison'][0] != df2['Comparison'][0]) & (df1['Subclass'][0] == df2['Subclass'][0]):\n",
    "                        df_all = df.loc[(df[\"Comparison\"] == comparison1) | (df[\"Comparison\"] == comparison2)]\n",
    "                        \n",
    "#                       default for merge is inner: use intersection of keys from both frames - find genes present in BOTH df, hence intersection\n",
    "                        df_intersection = df1.merge(df2.drop_duplicates(), on=['Gene'],\n",
    "                                    indicator=True)\n",
    "                        df_intersection = df_intersection.reset_index(drop=True)\n",
    "\n",
    "                        for gene in list(df_all['Gene']):\n",
    "                            #flags for all below\n",
    "                            brain_region = 2\n",
    "                            expression = -1000\n",
    "                            direction = \"specific to subclass\"\n",
    "                            regulation = \"NAN\"\n",
    "                            difference = \"specific to subclass\"\n",
    "                            ALM = False\n",
    "                            VISp = False\n",
    "                            CCA = float(list(df_all.loc[df_all['Gene'] == gene]['CCA'])[0])\n",
    "                            RCCA = float(list(df_all.loc[df_all['Gene'] == gene]['RCCA'])[0])\n",
    "                            RTSS = float(list(df_all.loc[df_all['Gene'] == gene]['RTSS'])[0])\n",
    "                            neuronalclass = f\"{str(df_intersection['Class_x'][0])}\"\n",
    "                            subclass = f\"{str(df_intersection['Subclass_x'][0])}\"\n",
    "                            conditions = \"Conditions Unlabeled\"\n",
    "\n",
    "                            if gene in list(df_intersection['Gene']):\n",
    "                                brain_region = 'ALM and VISp'\n",
    "                                \n",
    "#                             gene has to be in ALM and VISp since it is in intersection (here we find brain region origin of gene in DE test)\n",
    "                                if(df1['brain_region'][0] == 'ALM'):\n",
    "                                    ALMFC = float(df1.loc[df1[\"Gene\"] == gene]['avg_logFC'])\n",
    "                                elif(df1['brain_region'][0] == 'VISp'):\n",
    "                                    VISpFC = float(df1.loc[df1[\"Gene\"] == gene]['avg_logFC'])\n",
    "                                    \n",
    "                                if(df2['brain_region'][0] == 'ALM'):\n",
    "                                    ALMFC = float(df2.loc[df2[\"Gene\"] == gene]['avg_logFC'])\n",
    "                                elif(df2['brain_region'][0] == 'VISp'):\n",
    "                                    VISpFC = float(df2.loc[df2[\"Gene\"] == gene]['avg_logFC'])\n",
    "                                \n",
    "                                expression = (ALMFC + VISpFC)/2\n",
    "                                difference = ALMFC/VISpFC\n",
    "                                intersection = True\n",
    "                                \n",
    "                                if clin == 'y':\n",
    "                                    conditions = f\"{str(list(df_intersection.loc[df_intersection['Gene'] == gene]['Conditions_x'])[0])}\"\n",
    "\n",
    "                                if ((ALMFC <0) and (VISpFC <0)) or ((ALMFC >0) and (VISpFC >0)):\n",
    "                                    direction = \"same\"\n",
    "                                    regulation = signer(expression)\n",
    "\n",
    "#                               for genes that are DE in opposite directions (unconserved) between brain regions, collect in new df\n",
    "                                else:\n",
    "                                    diff1 = df1.loc[df1['Gene'] == gene]\n",
    "                                    diff2 = df2.loc[df2['Gene'] == gene]\n",
    "                \n",
    "                                    df_unconservedprep = pd.concat([diff1,diff2])\n",
    "                                    df_unconserved = pd.concat([df_unconserved,df_unconservedprep],ignore_index = True)\n",
    "                        \n",
    "                                    direction = \"different\"\n",
    "                                    regulation = \"differs between subclasses\"\n",
    "\n",
    "                            # gene wasn't DE in both ALM and VISp; let's find which brain region it was DE in\n",
    "\n",
    "                            elif ((gene in list(df1['Gene'])) & (df1['brain_region'][0] == 'ALM')):\n",
    "                                g1FC = float(df1.loc[df1[\"Gene\"] == gene]['avg_logFC'])\n",
    "                                expression = g1FC\n",
    "                                brain_region = 'ALM'\n",
    "                                regulation = signer(expression)\n",
    "                                \n",
    "                                if clin == 'y':\n",
    "                                    conditions = f\"{str(list(df1.loc[df1['Gene'] == gene]['Conditions'])[0])}\"\n",
    "\n",
    "                            elif ((gene in list(df1['Gene'])) & (df1['brain_region'][0] == 'VISp')):\n",
    "                                g1FC = float(df1.loc[df1[\"Gene\"] == gene]['avg_logFC'])\n",
    "                                expression = g1FC\n",
    "                                brain_region = 'VISp'\n",
    "                                regulation = signer(expression)\n",
    "\n",
    "                                if clin == 'y':                                \n",
    "                                    conditions = f\"{str(list(df1.loc[df1['Gene'] == gene]['Conditions'])[0])}\"\n",
    "\n",
    "\n",
    "                            elif ((gene in list(df2['Gene'])) & (df2['brain_region'][0] == 'VISp')):\n",
    "                                g2FC = float(df2.loc[df2[\"Gene\"] == gene]['avg_logFC'])\n",
    "                                expression = g2FC\n",
    "                                brain_region = 'VISp'\n",
    "                                regulation = signer(expression)\n",
    "                                if clin == 'y':\n",
    "                                    conditions = f\"{str(list(df2.loc[df2['Gene'] == gene]['Conditions'])[0])}\"\n",
    "\n",
    "\n",
    "                            elif ((gene in list(df2['Gene'])) & (df2['brain_region'][0] == 'ALM')):\n",
    "                                g2FC = float(df2.loc[df2[\"Gene\"] == gene]['avg_logFC'])\n",
    "                                expression = g2FC\n",
    "                                brain_region = 'ALM'\n",
    "                                regulation = signer(expression)\n",
    "                                if clin == 'y':\n",
    "                                    conditions = f\"{str(list(df2.loc[df2['Gene'] == gene]['Conditions'])[0])}\"\n",
    "                                \n",
    "                            if (brain_region == 'ALM'):\n",
    "                                rows.append([neuronalclass, subclass, gene, brain_region, conditions, CCA, RCCA, RTSS, expression, expression, 0, direction, regulation, difference])\n",
    "                            elif (brain_region == 'VISp'):\n",
    "                                rows.append([neuronalclass, subclass, gene, brain_region, conditions, CCA, RCCA, RTSS, expression, 0, expression, direction, regulation, difference])\n",
    "                            else:\n",
    "                                rows.append([neuronalclass, subclass, gene,  brain_region, conditions, CCA, RCCA, RTSS, expression, ALMFC, VISpFC, direction, regulation, difference])\n",
    "\n",
    "    df_DiffGenes = pd.DataFrame(rows, columns=['Class','Comparison', \"Gene\", \"brain_region\", \"Conditions\", \"CCA\", 'RCCA', 'RTSS', \"avg_logFC\", \"ALM logFC\", \"VISp logFC\", \"Direction\", \"Regulation (A vs B)\", \"% (ALM/VISp) logFC\"])\n",
    "    df_DiffGenes = df_DiffGenes.drop_duplicates(subset= ['Class','Comparison', \"Gene\", 'brain_region'], keep='first')\n",
    "    df_DiffGenes\n",
    "    \n",
    "    df_unconserved = df_unconserved.drop_duplicates(subset= ['Class','Comparison', \"Gene\", 'brain_region'], keep='first')\n",
    "\n",
    "\n",
    "    return GO, cellclass, brain_region, df_DiffGenes, df_unconserved"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subclass ID (SCID) Collapser\n",
    "requires df to first be passed through DE Collapser - takes a collapsed df (output from DE Collapser) and determines which genes define subclasses across brain regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SCID(prepped_df):\n",
    "    \n",
    "  #prep:\n",
    "    prepped_df['UpReg'], prepped_df['DownReg'], prepped_df['First'], prepped_df['Second'] = [0,0,0,0]\n",
    "               \n",
    "\n",
    "#           label columns for first and second subclass in DE test to determine up or down regulation\n",
    "    prepped_df[['First','Second']] = prepped_df['Comparison'].str.split(\"vs\", n = 1, expand = True) \n",
    "\n",
    "#           remove unecessary white spaces\n",
    "    prepped_df['First'] = prepped_df['First'].str.strip()\n",
    "    prepped_df['Second'] = prepped_df['Second'].str.strip()\n",
    "\n",
    "#           using idea that FC>0 means first subclass upreg and second subclass down and vice versa, label reg columns\n",
    "    prepped_df.loc[prepped_df['avg_logFC'] > 0, 'UpReg'] = prepped_df['First']\n",
    "    prepped_df.loc[prepped_df['avg_logFC'] > 0, 'DownReg'] = prepped_df['Second']\n",
    "\n",
    "    prepped_df.loc[prepped_df['avg_logFC'] < 0, 'UpReg'] = prepped_df['Second']\n",
    "    prepped_df.loc[prepped_df['avg_logFC'] < 0, 'DownReg'] = prepped_df['First']\n",
    "    \n",
    "    L2_3IT = ['L2/3 IT vs L5 IT', 'L2/3 IT vs L6 CT', 'L2/3 IT vs L5 PT']\n",
    "    L5IT = ['L2/3 IT vs L5 IT','L5 IT vs L5 PT','L5 IT vs L6 CT']\n",
    "    IT = ['L2/3 IT vs L5 PT', 'L2/3 IT vs L6 CT', 'L5 IT vs L5 PT','L5 IT vs L6 CT']\n",
    "    L5PT = ['L2/3 IT vs L5 PT','L5 IT vs L5 PT','L5 PT vs L6 CT']\n",
    "    L6CT = ['L2/3 IT vs L6 CT','L5 IT vs L6 CT','L5 PT vs L6 CT']\n",
    "    VIP = ['VIP vs PV', 'VIP vs SST']\n",
    "    SST = ['SST vs PV', 'VIP vs SST']\n",
    "    PV = ['SST vs PV', 'VIP vs PV']\n",
    "\n",
    "#       dictionary for all possible subclasses \n",
    "    SubclassDict = {'L2/3 IT': L2_3IT, 'L5 IT': L5IT, 'IT': IT, 'L5 PT': L5PT, 'L6 CT': L6CT, 'VIP': VIP, 'SST': SST, 'PV': PV}\n",
    "\n",
    "    for gene in prepped_df[\"Gene\"].unique():\n",
    "        \n",
    "        df_comparisons = prepped_df[prepped_df[\"Gene\"] == gene]\n",
    "        \n",
    "#       can add filter for min number of DE tests for given gene to be required for a subclass ID or can keep as is if interested in all combos\n",
    "        #solves issue of ASD/Neurodev vs Neurodev/ASD; compList is list of all comparisons involved in DE of a given gene\n",
    "        \n",
    "        compList = df_comparisons['ComparisonList'].unique()[0].split(\",\")\n",
    "              \n",
    "#       lists to add subclasses(subclass)/regulation(reg) groups when using DE collapser before Morpheus\n",
    "        subclass = []\n",
    "        Reg = []\n",
    "        \n",
    "        print(\"gene = \" + gene)\n",
    "                              \n",
    "#       identifies all subclasses present in subsetted dataset and sorts them\n",
    "        SubclassList = list(set(list(df_comparisons['First'].unique()) + list(df_comparisons['Second'].unique())))\n",
    "\n",
    "#       efficiency optimizing for groups such as IT; e.g. if glut classes are involved keep IT - otherwise remove  \n",
    "        if (set(['L2/3 IT','L5 IT']).issubset(set(SubclassList))):\n",
    "            SubclassList.append('IT')\n",
    "            \n",
    "        SubclassList = sorted(SubclassList)\n",
    "        \n",
    "        SubclassDEList = []\n",
    "        \n",
    "#       focus on only subclasses of subsetted data; sorting subclassList sorts SubclassDEList\n",
    "        for SC in SubclassList:\n",
    "            SubclassDEList.append(SubclassDict[SC])\n",
    "#             print(SubclassDEList)\n",
    "        \n",
    "        #loop through all potential subclass lists to check for presence of each subclass\n",
    "        for i in SubclassDEList:\n",
    "#             print(\"i equals: \" + str(i))\n",
    "\n",
    "            if (set(i).issubset(set(compList))):\n",
    "\n",
    "    #           make subset of DE tests for given subclass for a given gene\n",
    "                df_comparing = df_comparisons[df_comparisons['Comparison'].isin(i)].copy()\n",
    "\n",
    "    #                 check for conserved direction on group subclasses (e.g. IT)\n",
    "                if i == IT:\n",
    "                    df_comparing.loc[(df_comparing['UpReg'] == 'L2/3 IT') | (df_comparing['UpReg'] == 'L5 IT'),'UpReg'] = 'IT'\n",
    "                    df_comparing.loc[(df_comparing['DownReg'] == 'L2/3 IT') | (df_comparing['DownReg'] == 'L5 IT'), 'DownReg'] = 'IT'\n",
    "\n",
    "            #check for conserved direction on individual subclasses; does this by checking that only one unique subclass is down or up regulated\n",
    "                for SC in SubclassList:\n",
    "\n",
    "#                     print(SC)\n",
    "\n",
    "    #                       set(list(df_comparing['Down/UpReg']) should only have one subclass for the following to work (hence, why edited above for IT)\n",
    "                    if (set(list(df_comparing['DownReg'])).issubset(set([SC]))):\n",
    "\n",
    "                        print(\"SC downreg equals: \" + SC)\n",
    "\n",
    "                        subclass.append(df_comparing['DownReg'].unique()[0])\n",
    "                        Reg.append('DownReg')\n",
    "\n",
    "                    elif (set(list(df_comparing['UpReg'])).issubset(set([SC]))):\n",
    "\n",
    "                        print(\"SC upreg equals: \" + SC)\n",
    "\n",
    "                        subclass.append(df_comparing['UpReg'].unique()[0])\n",
    "                        Reg.append('UpReg')\n",
    "\n",
    "#       reg/subclass - xSC ySC / up down are paired\n",
    "       \n",
    "        prepped_df.loc[prepped_df['Gene'] == gene, 'Subclass'] = ';'.join(subclass)\n",
    "        prepped_df.loc[prepped_df['Gene'] == gene, 'Direction of Regulation'] = '/'.join(Reg)\n",
    "\n",
    "    return prepped_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CortexSCID (for use only when interested in cortex-specific SCID)\n",
    "Develops SCID to identify genes that define subclasses in one cortex by incorporating genes that were conserved as well as those exclusively DE in one cortex - addresses issue where failing to consider conserved genes when using SCID overlooks potential subclass labels when using SCID in cortex. \n",
    "\n",
    "\n",
    "Also removes genes where SCID labels are already apparent in conserved SCID (i.e. SCID conserved found gene X as upregulated in IT cells but SCID cortex flagged gene X as upregulated in IT and downregulated L5 PT; CortexSCID keeps gene X downregulated for L5 PT)\n",
    "\n",
    "conserved = SCID output for conserved data\n",
    "\n",
    "Cortex = SCID output for all ALM or VISp data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CortexSCID(conserved, Cortex):\n",
    "    \n",
    "#   finds genes that are present in conserved and ALM/VISP specific\n",
    "    overlapGenes = Cortex[['Gene']].drop_duplicates().merge(conserved[['Gene']].drop_duplicates(), how = 'inner' , indicator=False)\n",
    "    \n",
    "#   finds genes that are present in conserved and ALM/VISP specific with same ComparisonList and Subclass \n",
    "    overlapGenesSC = Cortex[['Gene','ComparisonList','Subclass']].drop_duplicates().merge(conserved[['Gene','ComparisonList','Subclass']].drop_duplicates(), how = 'inner' , indicator=False)\n",
    "\n",
    "    # special genes\n",
    "    UniqueSCID = set(overlapGenes.Gene).symmetric_difference(set(overlapGenesSC.Gene))\n",
    "    \n",
    "    print(overlapGenesSC)\n",
    "\n",
    "    # remove genes in ALM/VISp data that have are conserved and have same SCID in cortex and conserved\n",
    "    Cortex_unique = Cortex.loc[~Cortex['Gene'].isin(overlapGenesSC['Gene'])]\n",
    "\n",
    "    # deals with genes that are conserved but have a different SCID in cortex and conserved\n",
    "    UniqueSCID\n",
    "\n",
    "    for gene in UniqueSCID:\n",
    "        print(gene)\n",
    "        \n",
    "        cortexSC = Cortex_unique.loc[Cortex_unique['Gene'] == gene]['Subclass'].unique()[0].split(\";\")\n",
    "        conservedSC = conserved.loc[conserved['Gene'] == gene]['Subclass'].unique()[0].split(\";\")\n",
    "        DirectionCortexSC = Cortex_unique.loc[Cortex_unique['Gene'] == gene]['Direction of Regulation'].unique()[0].split(\"/\")\n",
    "\n",
    "        print(\"cortexSC = \" + str(cortexSC))\n",
    "        print(DirectionCortexSC)\n",
    "        print('conservedSC = ' + str(conservedSC))\n",
    "\n",
    "        if {''}.issubset(conservedSC):\n",
    "            conservedSC.remove('')\n",
    "\n",
    "        if {''}.issubset(cortexSC):\n",
    "            cortexSC.remove('')\n",
    "\n",
    "#       KeepSCID always pulls out genes with extra SCID labels that aren't in conserved SCID\n",
    "        KeepSCID = set(cortexSC).symmetric_difference(set(conservedSC))\n",
    "        KeepSCID = sorted(list(KeepSCID))\n",
    "\n",
    "    #     KeepSCID = ('/').join((KeepSCID))\n",
    "        print('Keep ' + str(KeepSCID))\n",
    "\n",
    "        DirectionList = []\n",
    "        \n",
    "#       appends directions with corresponding index of subclass; e.g. linked lists cortexSC = ['PV', 'VIP'] DirectionCortexSC = ['UpReg', 'DownReg']; assume we remove PV since it is flagged in conserved and cortex; thus want to keep ['VIP'] and ['Downreg']   \n",
    "        for SC in KeepSCID:\n",
    "            DirectionList.append(DirectionCortexSC[cortexSC.index(SC)])\n",
    "\n",
    "        print(\"Direction List = \" + str(DirectionList))\n",
    "\n",
    "        Cortex_unique.loc[Cortex_unique['Gene'] == gene, 'Subclass'] = ('/').join(KeepSCID)\n",
    "        Cortex_unique.loc[Cortex_unique['Gene'] == gene, 'Direction of Regulation'] = ('/').join(DirectionList)\n",
    "        \n",
    "        if(len(conservedSC)>=1):\n",
    "            Cortex_unique.loc[Cortex_unique['Gene'] == gene, 'Asterisk'] = '*'\n",
    "\n",
    "    return Cortex_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Signer\n",
    "pass a number - returns up or down regulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def signer(number):\n",
    "    if number > 0:\n",
    "        return \"up-regulation\"\n",
    "    elif number <0:\n",
    "        return \"down-regulation\"\n",
    "    else:\n",
    "        return \"reg 0 or not calculated\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MorphPrep \n",
    "\n",
    "Produces a tuple with 4 df_files (prepped_df, Matrix, Meta_genes, Meta_comparison) that upon saving as CSVs can be directly applied to Broad Institute app Morpheus\n",
    "\n",
    "relies on all above functions. Option to use DECollapser - if yes, SCID/cortexSCID is also an option\n",
    "\n",
    "df_clinPrep = df generated from ClinPrep\n",
    "\n",
    "df_combo = legacy combined dataset\n",
    "\n",
    "df_ClinList = manually made condition dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def MorphPrep(df_clinPrep, df_combo, df_ClinList):\n",
    "    \n",
    "    # clinical or not clinical\n",
    "    while True: \n",
    "        clinical = input(\"Interested in clinical analysis?(y/n)): \")\n",
    "        if clinical == 'y':\n",
    "            df2 = ClinLabel(df_clinPrep, df_combo,df_ClinList)\n",
    "            break\n",
    "        elif clinical == 'n':\n",
    "            df2 = df_combo\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "    \n",
    "#   subset alone or use collapser+subset    \n",
    "    while True: \n",
    "        collapse = input(\"Would you like to collapse dataset?(y/n) (if dataset is already collapsed, please type 'n'): \")\n",
    "        \n",
    "        if collapse == 'y':\n",
    "            collapser = DECollapser(df2)\n",
    "            df3 = collapser[3]\n",
    "            \n",
    "            cellclass = collapser[1]\n",
    "            \n",
    "#           prepare df for comparing for brain-region specific to conserved  \n",
    "            dfSCIDConserved = collapser[3].copy()\n",
    "            dfSCIDConserved = dfSCIDConserved.loc[(dfSCIDConserved['Direction'] == 'same')]\n",
    "            \n",
    "            D = {'same':'same', 'diff':'different', 'specific': 'specific to subclass'}\n",
    "\n",
    "            while True: \n",
    "                direction = input(\"Select subclasses with avglogFC in the same, different (diff), or specific to brain region direction (specific): \")\n",
    "                if (direction == 'diff'):\n",
    "                    df3 = collapser[4]\n",
    "                    break\n",
    "                \n",
    "                elif (direction == 'same' or direction == 'specific'):\n",
    "                    \n",
    "                    if (direction == 'same'):\n",
    "                        df3 = df3.loc[(df3['Direction'] == D[direction])]\n",
    "                    \n",
    "                    elif direction == 'specific':\n",
    "                        while True:\n",
    "                            region = input(\"ALM or VISp: \")\n",
    "                            \n",
    "#                          offer opportunity if planning to use SCID so to not lose genes that may have not been exclusively DE but still contribute to subclass ID   \n",
    "                            SCIDplan = input(\"Are you planning to ID subclass specific genes on your collapsed brain region specific dataset?(y/n) \")\n",
    "                            \n",
    "                            if (region == 'ALM' and SCIDplan == 'y') or (region == 'VISp' and SCIDplan == 'y'):\n",
    "                                df3 = df3.loc[(df3['brain_region'] == region) | (df3['brain_region'] == 'ALM and VISp')]\n",
    "                                \n",
    "#                               to make downstream avg_logFC values consistent with region selected we change average to brain region log_FC; checked splice warning - no error  \n",
    "                                df3['avg_logFC'] = df3[f'{region} logFC']\n",
    "                                break\n",
    "                            \n",
    "                            elif (region == 'ALM' and SCIDplan == 'n') or (region == 'VISp' and SCIDplan == 'n'):\n",
    "                                df3 = df3.loc[(df3['brain_region'] == region)]\n",
    "                                break\n",
    "    \n",
    "                            else:\n",
    "                                print(\"Please enter valid input\")\n",
    "                    break\n",
    "                else:\n",
    "                    print(\"Please enter valid input\")\n",
    "            break\n",
    "        \n",
    "        elif collapse == 'n':\n",
    "            df3 = subsetter(df2)[3]\n",
    "            break\n",
    "        else:\n",
    "            print(\"Please enter valid input\")\n",
    "    \n",
    "    # generate matrix for values:   \n",
    "    prepped_df = df3.copy()\n",
    "    \n",
    "    # make meta_genes sorted as desired (\"order\")\n",
    "    prepped_df['Frequency'] = prepped_df.groupby('Gene')['Gene'].transform('count')\n",
    "    prepped_df['ComparisonList'] = 'NAN'\n",
    "    \n",
    "\n",
    "    for gene in prepped_df[\"Gene\"].unique():\n",
    "        df_comparisons = prepped_df[prepped_df[\"Gene\"] == gene]\n",
    "\n",
    "        #solves issue of ASD/Neurodev vs Neurodev/ASD; compList is list of all comparisons involved in DE of a given gene\n",
    "        compList = sorted(df_comparisons['Comparison'].tolist())\n",
    "        my_string = ','.join(compList)\n",
    "        prepped_df.loc[prepped_df['Gene'] == gene, \"ComparisonList\"] = my_string\n",
    "        \n",
    "        \n",
    "    #   repeat for dfSCIDConserved\n",
    "    if (collapse == 'y'):\n",
    "        if (direction == 'specific'):\n",
    "            if SCIDplan == 'y':\n",
    "\n",
    "                dfSCIDConserved['Frequency'] = dfSCIDConserved.groupby('Gene')['Gene'].transform('count')\n",
    "                dfSCIDConserved['ComparisonList'] = 'NAN'\n",
    "\n",
    "                for gene in dfSCIDConserved[\"Gene\"].unique():\n",
    "                    df_comparisons = dfSCIDConserved[dfSCIDConserved[\"Gene\"] == gene]\n",
    "\n",
    "                    #solves issue of ASD/Neurodev vs Neurodev/ASD; compList is list of all comparisons involved in DE of a given gene\n",
    "                    compList = sorted(df_comparisons['Comparison'].tolist())\n",
    "                    my_string = ','.join(compList)\n",
    "                    dfSCIDConserved.loc[dfSCIDConserved['Gene'] == gene, \"ComparisonList\"] = my_string\n",
    "\n",
    "#   SCID or not SCID\n",
    "    if (collapse == 'y'):\n",
    "        if (direction != 'diff'):\n",
    "            while True: \n",
    "                    ID = input(\"Would you like to ID subclass specific genes on your collapsed dataset?(y/n) \")\n",
    "                    if ID == 'y':\n",
    "\n",
    "                        prepped_df = SCID(prepped_df)                                                \n",
    "\n",
    "                    # if looking for motifs specific to brain regions, must take out class-relevant genes present across both brain regions and only keep those in which there is additional region-specific class info  \n",
    "                        if direction == 'specific':\n",
    "\n",
    "                            if SCIDplan == 'y':\n",
    "\n",
    "                                dfSCIDConserved = SCID(dfSCIDConserved)\n",
    "                                prepped_df = CortexSCID(dfSCIDConserved,prepped_df)                            \n",
    "\n",
    "                        break\n",
    "                    elif ID == 'n':\n",
    "                        break\n",
    "                    else:\n",
    "                        print(\"Please enter valid input\")\n",
    "                                 \n",
    "    prepped_df['Subclass'] = prepped_df['Subclass'].str.replace(\";\",\"/\")\n",
    "\n",
    "    prepped_df['ComparisonList'] = prepped_df['ComparisonList'].str.replace(',','/')\n",
    "    \n",
    "#   corrects diamond key - assures first subclass listed upregulates gene and second subclass listed downregulates gene \n",
    "    if (collapse == 'y'):\n",
    "        if (direction != 'diff'):\n",
    "            if (cellclass != 'all'):\n",
    "                prepped_df.loc[prepped_df['Direction of Regulation'] == 'DownReg/UpReg', 'Subclass'] = prepped_df.loc[prepped_df['Direction of Regulation'] == 'DownReg/UpReg']['Subclass'].apply(lambda x: ((',').join(x.split('/')[::-1])).replace(',','/'))\n",
    "                prepped_df.loc[prepped_df['Direction of Regulation'] == 'DownReg/UpReg', 'Direction of Regulation'] = prepped_df.loc[prepped_df['Direction of Regulation'] == 'DownReg/UpReg']['Direction of Regulation'].apply(lambda x: ((',').join(x.split('/')[::-1])).replace(',','/'))\n",
    "    \n",
    "    df_matrix = prepped_df[['Gene','Comparison','avg_logFC']]\n",
    "    pivot = df_matrix.pivot('Gene', 'Comparison','avg_logFC')\n",
    "    Matrix = pivot\n",
    "    \n",
    "    # generate matrix for gene annotations:    \n",
    "    if clinical == 'y':\n",
    "        \n",
    "#       pre-req for ID  \n",
    "        if (collapse == 'y'):\n",
    "        \n",
    "            if (direction != 'diff'):\n",
    "            \n",
    "                if ID == 'y':\n",
    "                    prepped_df.sort_values(['Conditions', 'Subclass', 'Frequency','ComparisonList', 'Direction of Regulation','Gene'], inplace=True, ascending=[True,False,False,True, False, True])\n",
    "                    prepped_df['Order'] = range(1, 1+len(prepped_df))\n",
    "\n",
    "                    df_genes = prepped_df[['Gene','ComparisonList','CCA', 'RCCA','RTSS','Frequency','Subclass','Direction of Regulation','Conditions', 'Order']]\n",
    "        \n",
    "        else:\n",
    "            prepped_df.sort_values(['Conditions', 'Frequency','ComparisonList','Gene'], inplace=True, ascending=[True,False,True,True])\n",
    "            prepped_df['Order'] = range(1, 1+len(prepped_df))\n",
    "            df_genes = prepped_df[['Gene','ComparisonList', 'Frequency','CCA', 'RCCA','RTSS','Conditions', 'Order']]\n",
    "        \n",
    "#       checked splice warning, not an issue\n",
    "        df_genes['Conditions'] = df_genes['Conditions'].str.replace(',','/')\n",
    "    \n",
    "    else:\n",
    "         if (collapse == 'y'):\n",
    "            if (direction != 'diff'):\n",
    "                if ID == 'y':\n",
    "                    prepped_df.sort_values(['Subclass','ComparisonList','Frequency', 'Direction of Regulation','Gene'], inplace=True, ascending=[True,False,False, False, True])\n",
    "                    prepped_df['Order'] = range(1, 1+len(prepped_df))\n",
    "                    df_genes = prepped_df[['Gene','ComparisonList','CCA', 'RCCA','RTSS','Frequency','Subclass','Direction of Regulation','Order']]\n",
    "        \n",
    "         else:\n",
    "            prepped_df['Order'] = range(1, 1+len(prepped_df))\n",
    "            prepped_df.sort_values(['Frequency','ComparisonList','Gene'], inplace=True, ascending=[False,True,True])\n",
    "            df_genes = prepped_df[['Gene','ComparisonList', 'Frequency','CCA', 'RCCA','RTSS', 'Order']]\n",
    "  \n",
    "    if (collapse == 'y'):\n",
    "        if (direction == 'diff'):\n",
    "            df_genes = prepped_df[['Gene','ComparisonList', 'Frequency','CCA', 'RCCA','RTSS']]      \n",
    "        \n",
    "#   in order to prepare for Morpheus, we drop redundant gene rows; checked splice warning, not an issue   \n",
    "    df_genes.drop_duplicates(subset=['Gene'], keep=\"first\", inplace = True)\n",
    "\n",
    "    # generate matrix for comparison annotations: (change to prepped_df)\n",
    "    df_comparison = prepped_df[['Comparison','brain_region','Class', 'Subclass']]\n",
    "    Meta_comparison = df_comparison\n",
    "    \n",
    "    Meta_genes = df_genes        \n",
    "\n",
    "    return prepped_df, Matrix, Meta_genes, Meta_comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Morpheus creator testing terminal\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# to run, change file paths to locations on your computer. For questions concerning each file, review readme\n",
    "\n",
    "df_clin= pd.read_csv(r'C:\\SFARI.csv')\n",
    "\n",
    "df_clin = clinExploder(df_clin)\n",
    "\n",
    "df_DEGenes = pd.read_csv(r'C:\\DE Gene List.csv')\n",
    "\n",
    "df_clinPrep = ClinPrep(df_clin,df_DEGenes)\n",
    "\n",
    "df_combo = pd.read_csv(r'C:\\Combined_filtered_GO.csv')\n",
    "\n",
    "df_ClinList = pd.read_csv(r'C:\\Narrowed Condition List.csv')\n",
    "\n",
    "MorphFiles = MorphPrep(df_clinPrep, df_combo, df_ClinList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# view ouput df from applying tools\n",
    "analyzed_df = MorphFiles[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Morpehus saver testing terminal - save files for uploading to Morpheus and visualizing\n",
    "\n",
    "pivot = MorphFiles[1]\n",
    "\n",
    "df_genes = MorphFiles[2]\n",
    "\n",
    "df_comparison = MorphFiles[3]\n",
    "\n",
    "pivot.to_csv(r'C:\\Matrix.csv')\n",
    "\n",
    "df_genes.to_csv(r'C:\\Meta_genes.csv', index = False)\n",
    "\n",
    "df_comparison.to_csv(r'C:\\Meta_comparison.csv', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
